noir-projects/aztec-nr/aztec/src/contract_self.nr

```diff
@@ -0,0 +1,239 @@
+use crate::{
+    context::{private_context::PrivateContext, public_context::PublicContext},
+    event::{event_emission::emit_event_in_private, event_interface::EventInterface},
+};
+use super::{context::utility_context::UtilityContext, event::event_emission::emit_event_in_public};
+use protocol_types::{
+    address::AztecAddress, constants::NULL_MSG_SENDER_CONTRACT_ADDRESS, traits::Serialize,
+};
+
+/// `ContractSelf` is the core interface for interacting with an Aztec contract's own state and context.
+///
+/// This struct is automatically injected into every contract function by the Aztec macro system and is
+/// accessible through the `self` variable. It provides a unified interface for accessing:
+/// - The contract's own address
+/// - The contract's storage
+/// - The execution context (private, public, or utility)
+///
+/// # How It's Injected
+///
+/// When you write a contract function with the `#[external("private")]`, `#[external("public")]`, or
+/// `#[external("utility")]` attributes, the Aztec macro system automatically transforms your function
+/// to include a `self` parameter of type `ContractSelf`.
+///
+/// For example, when you write:
+/// ```noir
+/// #[external("private")]
+/// fn transfer(amount: u128, recipient: AztecAddress) {
+///     // Your code here
+/// }
+/// ```
+///
+/// The macro transforms it to something like:
+/// ```noir
+/// fn transfer(inputs: PrivateContextInputs, amount: u128, recipient: AztecAddress) -> PrivateCircuitPublicInputs {
+///     let mut self = {
+///         let mut context = PrivateContext::new(inputs, args_hash);
+///         let storage = Storage::init(&mut context);
+///         ContractSelf::new_private(&mut context, storage)
+///     };
+///     // Your original code executes here with `self` available
+///     // ...
+///     self.context.finish()
+/// }
+/// ```
+///
```

Comment:
This is entirely unnecessary as user docs. They don't care about this at all.

Reply:
> (i'm being free with harsh language since you didn't write this yourself 😛)

I instructed AI to put this there as I felt devs might be curious about it so crying anyway 😭 😆

But agree it's better to just have this on the `comptime` macro functions so I dropped it in https://github.com/AztecProtocol/aztec-packages/pull/17884/commits/e822875780f5d1a0375c767b7a5a833c7649f485